classes/userGroup/UserGroup.php

```diff
@@ -134,12 +137,11 @@ public function getPermitSelfRegistration()
 
     /**
      * Set the "permit self-registration" flag
-     *
-     * @param bool $permitSelfRegistration
      */
-    public function setPermitSelfRegistration($permitSelfRegistration)
+    public function setPermitSelfRegistration(?bool $permitSelfRegistration)
```

Comment:
Is the `null` state necessary? It looks to me like this is always `true` or `false`.

Reply:
You are right. the `$permitSelfRegistration` should not be null.

I changed that in order to "fix" cases where there where a code segment that tried to set `$permitSelfRegistration` to `null`. Instead of adding the check when the `setPermitSelfRegistration` is called, I added this business logic to the actual `setPermitSelfRegistration` function. 

I am wondering whether, if changing the `?bool` to `bool`, the code will fail if the code will try to pass null value there. Also I am wondering how the `->default` value of a column will be filled in the DB if the object has a value of `null` for the specified param, or if the object has no value for the specified param. Will the default db value be applied to both of those cases?

Reply:
Yes, passing a `null` to a `bool`-typehinted parameter will fail; this is why we can't just batch-add type hints to classes without expecting a lot of breakage. I think we should be getting stricter comprehensively and consistently -- business logic, database, storage code, everything. The database definition of columns is usually pretty accurate (e.g. `permit_self_registration` does not permit `null` data, which is good).

If the PHP object has a `null` value, then the `INSERT` statement will pass the `NULL` along to the database. Inserting a `NULL` value in a `NOT NULL` column will cause an error, as it should. So when creating a new `UserGroup` object, the calling code will be responsible for setting an initial value of `true` or `false` for `setPermitSelfRegistration`. But that's not new -- it's always been the case.

Reply:
Changed